gluten-core/src/main/scala/io/glutenproject/extension/ExpandFallbackPolicy.scala

```diff
@@ -67,6 +67,45 @@ import org.apache.spark.sql.execution.exchange.{Exchange, ReusedExchangeExec}
 case class ExpandFallbackPolicy(isAdaptiveContext: Boolean, originalPlan: SparkPlan)
   extends Rule[SparkPlan] {
 
+  private def ignoreOneColumnarToRow(plan: SparkPlan): Boolean = {
+    // For native file scan, there is a chance that we can totally fallback to
+    // vanilla Spark file scan, but for a materialized `ColumnarShuffleExchangeExec`
+    // there must be a `ColumnarToRowExec` if we decide to fallback.
+    val hasScan = plan.collectLeaves().exists {
+      case _: BasicScanExecTransformer => true
+      case _ => false
+    }
+
+    // spotless:off
+    // It has no meaning to add ColumnarToRow eagerly if we will add it finally.
+    // So we ignore the ColumnarToRow to avoid fallback eagerly.
+    // For example: 1 is better than 2
+    //
+    // 1. We first run native operator then add ColumnarToRow
+    //      ColumnarExchange
+    //            |
+    //  HashAggregateTransformer
+    //            |
+    //      ColumnarToRow
+    //
+    // 2. We first add ColumnarToRow then run vanilla Spark operator
+    //     ColumnarExchange
+    //            |
+    //      ColumnarToRow
+    //            |
+    //      HashAggregate
+    //
+    // spotless:on
+    var numColumnarToRow = 0
+    var numRowToColumnar = 0
+    plan.foreach {
+      case _: ColumnarToRowExec => numColumnarToRow += 1
+      case _: RowToColumnarExec => numRowToColumnar += 1
+      case _ =>
+    }
+    !hasScan && numColumnarToRow == 1 && numRowToColumnar == 0
```

Comment:
I think it may be more straightforward to collect all leaf nodes that are query stage nodes (they represent last stages). And count all such leaf nodes who have columnar output to get `requiredColumnarToRowWhenFallbackWholeStage`. Then, we get

```
netFallbacks = fallbacks -  requiredColumnarToRowWhenFallbackWholeStage
```
We may just need to check whether `netFallbacks` meets user-specified `fallbackThreshold`.
For your case where `fallbackThreshold = 1`,  we can find `fallbacks = 1`, but `requiredColumnarToRowWhenFallbackWholeStage = 1`, so the fallback policy will not take effect for this case. 

Using `netFallbacks` looks reasonable also for other threshold settings, no matter whether this stage has scan node. Do you think so? Thanks!

Reply:
In one stage, there may be one more ColumnarToRow for one leaf node, for example:
```
  ColumnarExchange
         |
HashAggregateTransformer
         |
   ColumnarToRow
         |
      Project
         |
   ColumnarToRow
         |
 ProjectTransformer
         |
   ColumnarToRow
         | 
   HashAggregate
         |
        ... 
```
That's why I just improve the case that fallback threshold = 1. Things will be complex if the threshold is bigger than 1.

Reply:
Let's merge this version firstly. I can try to do some optimization in a separate PR. Thanks!

Reply:
thank you @PHILO-HE !